xdsl/dialects/riscv.py

```diff
@@ -127,15 +138,18 @@ class Register:
         "ft10": 30,
         "ft11": 31,
     }
+    ABI_INDEX_BY_NAME = RV32F_INDEX_BY_NAME
 
 
-@irdl_attr_definition
-class RegisterType(Data[Register], TypeAttribute):
+RegisterT = TypeVar("RegisterT", bound="Register")
+
+
+class RegisterType(GenericData[RegisterT], TypeAttribute):
     """
     A RISC-V register type.
     """
 
-    name = "riscv.reg"
+    register_class: type[RegisterT]
```

Comment:
I don't think we need this tbh, let's start with a less flexible non-generic attribute with two subclasses, and use abstract methods rather than do full generic approach.